packages/ckb-indexer/src/collector.ts

```diff
@@ -43,82 +43,117 @@ export class CKBCellCollector implements BaseCellCollector {
     }
-    if (this.queries.toBlock) {
-      utils.assertHexadecimal("toBlock", this.queries.toBlock);
+    if (queries.toBlock) {
+      utils.assertHexadecimal("toBlock", queries.toBlock);
     }
-    if (this.queries.order !== "asc" && this.queries.order !== "desc") {
+    if (queries.order !== "asc" && queries.order !== "desc") {
       throw new Error("Order must be either asc or desc!");
     }
-    if (this.queries.outputCapacityRange) {
+    if (queries.outputCapacityRange) {
       utils.assertHexadecimal(
         "outputCapacityRange[0]",
-        this.queries.outputCapacityRange[0]
+        queries.outputCapacityRange[0]
       );
       utils.assertHexadecimal(
         "outputCapacityRange[1]",
-        this.queries.outputCapacityRange[1]
+        queries.outputCapacityRange[1]
       );
     }
 
-    if (this.queries.outputDataLenRange) {
+    if (queries.outputDataLenRange) {
       utils.assertHexadecimal(
         "outputDataLenRange[0]",
-        this.queries.outputDataLenRange[0]
+        queries.outputDataLenRange[0]
       );
       utils.assertHexadecimal(
         "outputDataLenRange[1]",
-        this.queries.outputDataLenRange[1]
+        queries.outputDataLenRange[1]
       );
     }
 
-    if (this.queries.skip && typeof this.queries.skip !== "number") {
+    if (queries.outputDataLenRange && queries.data && queries.data !== "any") {
+      const dataLen = getHexStringBytes(queries.data);
+      if (
+        dataLen < Number(queries.outputDataLenRange[0]) ||
+        dataLen >= Number(queries.outputDataLenRange[1])
+      ) {
+        throw new Error("data length not match outputDataLenRange");
+      }
+    }
+
+    if (queries.skip && typeof queries.skip !== "number") {
       throw new Error("skip must be a number!");
     }
 
-    if (
-      this.queries.bufferSize &&
-      typeof this.queries.bufferSize !== "number"
-    ) {
+    if (queries.bufferSize && typeof queries.bufferSize !== "number") {
       throw new Error("bufferSize must be a number!");
     }
   }
 
+  public convertParams(): void {
+    // unWrap `ScriptWrapper` into `Script`.
+    if (this.queries.lock) {
+      if (instanceOfScriptWrapper(this.queries.lock)) {
+        validators.ValidateScript(this.queries.lock.script);
+        this.queries.lock = this.queries.lock.script;
+      }
+    }
+
+    // unWrap `ScriptWrapper` into `Script`.
+    if (this.queries.type && this.queries.type !== "empty") {
+      if (
+        typeof this.queries.type === "object" &&
+        instanceOfScriptWrapper(this.queries.type)
+      ) {
+        this.queries.type = this.queries.type.script;
+      }
+    }
+
+    if (!this.queries.outputDataLenRange) {
+      if (this.queries.data && this.queries.data !== "any") {
+        const dataLenRange = getHexStringBytes(this.queries.data);
+        this.queries.outputDataLenRange = [
+          "0x" + dataLenRange.toString(16),
+          "0x" + (dataLenRange + 1).toString(16),
+        ];
+      }
+    }
```

Comment:
we should also check if both `data` and `outputDataLenRange` are present in the `QueryOption`

```ts
// pseudocode
if (outputDataLenRange && data) {
  asserts(
    byteLengthOf(data) >= outputDataLenRange[0] &&
      byteLengthOf(data) < outputDataLenRange[1],
    throwError("mismatched datalen and range ...")
  );
}
```

Reply:
validate logic has been done in validation function  https://github.com/nervosnetwork/lumos/pull/318/files#diff-b6c230b5ef87aecafc6364827f77b3983aa96df957e497617063ca55a072fe92R108

and I add test at [next commit](https://github.com/nervosnetwork/lumos/pull/318/commits/21f80dc7535ca0277e741479e423d793dbd32168)